### ISSUE TITLE
Allow specifying "equal" function for signal based inputs.

### ISSUE THREAD

**Issue by eiswind** [angular/angular#54111]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

Currently we have a lodash deep equals on our signals that are bound to an input setter like

```typescript
 _message = signal({} as Message, { equal: isEqual });
@Input({ required: true }) set message(value: Message) {
  this._message.set(value);
}
```
We could not find a way to do the same with a signal based input.

### Proposed solution

Allow setting an equal function on signal based inputs.

### Alternatives considered

Maybe we are getting this whole thing wrong?

**Comment by eiswind** [comment]
@JoostK sorry for opening this twice, I could not find it with the search function, so I thought I didn't hit submit last week. Is there anything I can do to clear this up / help with clarifocation?

**Comment by lacolaco** [comment]
I have the same request for `model()`. It is needed for implementing custom form control for non-primitive value with `model()`. Below is an example; 

```ts
export type Time = {
  hour: number;
  minute: number;
};

export class TimeInputComponent implements ControlValueAccessor {
  value = model<Time>({ hour: 0, minute: 0 });

  #onChangeListener = (_: Time) => {};

  constructor() {
    effect(() => {
      this.#onChangeListener(this.value());
    });
  }

  // ControlValueAccessor implementation
  writeValue(value: Time): void {
    this.value.set(value);
  }

  registerOnChange(fn: (v: Time) => void): void {
    this.#onChangeListener = fn;
  }
}
```

The `value` model input doesn't have a custom `equal` function for `Time` type. So every `set()` occurs the effect which emits change to the bound form control. And the current workaround is this:

```
export function isEqualTime(a: Time, b: Time) {
  return a.hour === b.hour && a.minute === b.minute;
}

export class TimeInputComponent implements ControlValueAccessor {
  readonly value = model<Time>({ hour: 0, minute: 0 });
  readonly #valueChange = computed(() => this.value(), { equal: isEqualTime }); // <-- distinct changed value

  #onChangeListener = (_: Time) => {};

  constructor() {
    effect(() => {
      this.#onChangeListener(this.#valueChange());
    });
  }

  // ControlValueAccessor implementation
  writeValue(value: Time): void {
    this.value.set(value);
  }

  registerOnChange(fn: (v: Time) => void): void {
    this.#onChangeListener = fn;
  }
}
```

**Comment by ptandler** [comment]
I agree, that `equal` should be added to input signal options, we also use it a lot (via setter currently)

**Comment by JeanMeche** [comment]
I think this has already been stated before, but the equal function currently comes with limitations, namely when passing down to inputs. 

https://github.com/angular/angular/blob/a84cb62522bff516e3b05eaa7b7ff9a0ca46c561/packages/core/src/render3/component_ref.ts#L352-L362

If the reference doesn't change then inputs won't get updated.

**Comment by pkozlowski-opensource** [comment]
## tl;dr;

we are _not_ planning on adding the `equals` option to inputs as this is not feasible with the current design of the change detection and inputs. There are use-cases that could be improved with other APIs, though. Capturing our thinking / discussion so far below.

## Details

When we consider the question of inputs and the equality there are 2 different use-cases to be taken into account:
1. making two referentially equal (===) bindings to be consider as "different" and cause write to an input;
2. have the input consider two referentially unequal (!==) bindings as "the same"

### Equal bindings considered unequal

Consider the following binding: `<user-details [user]="userDetailExpression" />`. When the `userDetailExpression` represents an object (ex.: `{first: "John", last: "Doe"}`) it is tempting to _mutate_ this object and _expect_ the binding to be re-evaluated. But this is not how the change detection works - a binding uses === / Object.is and will _not_ write to an input (signals or not) when the reference is the same.

We can't do much about the case (1) - without rethinking change detection, immutability etc. This is _not_ likely to happen. 

### Unequal bindings considered equal 

Continuing with our example binding `<user-details [user]="userDetailExpression" />` it might happen that a reference to the `user` object changes but the field values are the same. In this case a directive / component might want to consider such references as "equal". This seems to be a valid use-case and there are multiple ways of thinking about it:
* actually add the equality to `input` for this particular use-case;
* expose a previous value to `computed` and allow people to deal with those situations on a computed level (see a related discussion in https://github.com/angular/angular/issues/54339)

Again, this use-case seems valid but requires more thinking / design.

**Comment by asherwoodcnet** [comment]
> Adding the `equality` option to the `input` might be confusing to users since they might expect the first case presented here to be taken into account.

This reasoning makes perfect sense to me. Could we get the functionality, but avoid the confusion by using a name that would convey "this will only filter potential changes"? `isUnchanged` comes to mind.

**Comment by ptandler** [comment]
> > Adding the `equality` option to the `input` might be confusing to users since they might expect the first case presented here to be taken into account.
> 
> This reasoning makes perfect sense to me. Could we get the functionality, but avoid the confusion by using a name that would convey "this will only filter potential changes"? `isUnchanged` comes to mind.

You mean something like this?

```ts
@Component({
    selector: 'user-details',
})
class UserDetails {
   user = input.required<User>();
   // use this to get notificationy only when changed
   private readonly currentUser = this.user.isUnchanged(isEqual); // maybe equal function is optional and uses isEqual by default
}
```

or even shorter:

```ts
@Component({
    selector: 'user-details',
})
class UserDetails {
   user = input.required<User>().isUnchanged(isEqual);
}
```

**Comment by asherwoodcnet** [comment]
Wow, I like that idea a lot! I was actually thinking of something like:
```ts
class UserDetails {
   user = input.required<User>({ isUnchanged: (prev, next) => ... });
```
But your idea is cleaner, IMO. maybe `.withDiffer` to match the existing angular array/object diff paradigm?
```ts
class UserDetails {
   user = input.required<User>().withDiffer(isEqual);
```

**Comment by rdamazio** [comment]
Can this at least be considered for `model<>`, which tends to be used in more complex use cases than input?

e.g.

```ts
model.update((val: Foo): Foo => {
  val.x++;
  return val;
});
```

currently does not emit, which is itself very counterintuitive.

**Comment by asherwoodcnet** [comment]
kind of off topic, but @rdamazio I would normally do something more like this:
```ts
model.update(val => ({
  ...val,
  x: val.x + 1
}));
```
which circumvents the equality problem

**Comment by rdamazio** [comment]
> kind of off topic, but [@rdamazio](https://github.com/rdamazio) I would normally do something more like this:
> 
> model.update(val => ({
>   ...val,
>   x: val.x + 1
> }));
> which circumvents the equality problem

Thanks, that's a good hint, I'll use it in some places.
Unfortunately it may also involve a lot of copying, if val is large (arguably that'd also make a simple deep equality comparison expensive, I know, but with custom comparators you can get smart about it).

**Comment by ShacharHarshuv** [comment]
> Adding the equality option to the input might be confusing to users since they might expect the first case presented here to be taken into account.

That's a weak excuse, since the same exact reasoning can be applied to `signal` and `computed`, yet we still have an `equal` option. 

```ts
const x = computed(() => ...);
effect(() => {
  console.log(x());
})
x().name = 'Shahar'; // is anyone going be surprised the new value is not logged in this case? Why would it be any different for inputs? 
```

I don't see any reason why `signal` and `computed` should have an `equal` option but `input` should not. There is no reason to rename `equal` to `isUnchanged` either since it just creates inconsistency in the Angular api design. 

> But your idea is cleaner, IMO....

No, it's not, because it requires two members (`user` and `currentUser`) where the sensible `equal` solution only has one.

**Comment by asherwoodcnet** [comment]
I totally forgot about this ticket! after having sat with signals for over a year, I'm inclined to agree that an input option makes the most sense. for what it's worth, I haven't felt the pain of this missing equality function, since I'm usually splitting object inputs into computed signals anyway. I still think it's a good idea though!

> it requires two members (user and currentUser) where the sensible equal solution only has one.

I don't think that's accurate, since this would be possible:
```ts
user = input.required<User>().withDiffer(isEqual);
```
but either way, I've soured on the idea of accomplishing this with method chaining. it would feel odd to me to have a convenience method that exists *only* on inputs that is essentially just `(sourceSignal, equalFn) => computed(() => sourceSignal(), { equal: equalFn })`.

**Comment by jsangmeister** [comment]
This is especially a problem for custom controls in signal-based forms with more complex values. If I want to use a cuomt control with e.g. `value = model(new Set<string>())`, I cannot provide a custom equality function and have to manually call
```
form.customControl().value.update((x) => {
  const newSet = new Set(x);
  newSet.add(newValue);
  return newSet;
});
```
everytime I want to update the value from the outside. This is tedious and error prone.
